File: pontifex/pontifex.rb
# This program implements the Solitaire encryption algorithm 
# as described in the appendix of the book Cryptonomicon by 
# Neal Stephenson:

# https://www.schneier.com/academic/solitaire/

# Solitaire is a cypher that can be performed using a deck of 
# playing cards. If two people each have a deck of playing 
# cards in the same order, each can generate a keystream that 
# can be used to encypher and decypher a message.

class Integer
  def charValue() 
    value = self
    while value < 1
      value += 26
    end
    value = (value - 1) % 26 + 1
    return (value + 64).chr
  end
end

class String
  def intValue() 
    return self.ord - 64
  end
  
  def plus(c)
    return (self.intValue + c.intValue).charValue
  end
  
  def minus(c)
    return (self.intValue - c.intValue).charValue
  end
  
  def add(s)
    return self.split("").zip(s.split(""))
      .map { |pair| pair[0].plus(pair[1]) }
      .join
  end
  
  def subtract(s)
    return self.split("").zip(s.split(""))
      .map { |pair| pair[0].minus(pair[1]) }
      .join
  end
  
  def blocks() 
    return self.scan(/.{5}/).join(" ")
  end
end

class Card
  attr_accessor :suit
  attr_accessor :rank
  
  def initialize(suit, rank)
    @suit = suit
    @rank = rank
  end
  
  def to_s
    return rank_str() + suit_str()
  end  

  def rank_str()
    if suit == :joker
      return rank == 1 ? "A" : "B"
    else
      case rank
      when 1
       return "A"
      when 11
        return "J"
      when 12
        return "Q"
      when 13
        return "K"
      else
        return rank.to_s
      end
    end
  end
    
  def suit_str()
    case suit
    when :clubs
      return "♣️"
    when :diamonds
      return "♦️"
    when :hearts
      return "♥️"
    when :spades
      return "♠️"
    else
      return "🃏"
    end
  end

  def number() 
    case suit
    when :clubs
      return rank
    when :diamonds
      return rank + 13
    when :hearts
      return rank + 26
    when :spades
      return rank + 39
    else 
      return 53
    end
  end

  def letter
    case suit
    when :clubs, :hearts
      return rank.charValue
    when :diamonds, :spades
      return (rank + 13).charValue
    when :joker
      return nil
    end
  end
end 
  
class Deck
  attr_accessor :cards
  attr_accessor :jokerA
  attr_accessor :jokerB

  def initialize()
    @cards = []
    [:clubs, :diamonds, :hearts, :spades].each do |suit|
      13.times do |rank|
        cards.push(Card.new(suit, rank + 1))
      end
    end
    @jokerA = Card.new(:joker, 1)
    @jokerB = Card.new(:joker, 2)
    cards.push(jokerA)
    cards.push(jokerB)
  end
  
  def move(card, downBy) 
    index = cards.index(card)
    cards.delete(card)
    newIndex = (index + downBy - 1) % cards.count + 1
    cards.insert(newIndex, card)
  end
  
  def cut(count)
    @cards = cards[count...53] + 
             cards[0...count] +
             cards[53...54]
  end

  def tripleCut(card1, card2)
    index1 = cards.index(card1)
    index2 = cards.index(card2)
    if index1 > index2
      index1, index2 = index2, index1
    end
    @cards = cards[(index2 + 1)...cards.count] + 
             cards[index1...(index2 + 1)] +
             cards[0, index1]
    end
end 

class Generator
  def keystream(length)
    stream = ""
    while stream.length < length
      stream += more()
    end
    return stream[0..(length - 1)]
  end
  
  def more() 
    return "A"
  end  
end

class Example < Generator
  def more() 
    return "KDWUPONOWT"
  end  
end

class Solitaire < Generator
  attr_accessor :deck
  
  def initialize(passphrase)
    @deck = Deck.new
    
    if passphrase != nil 
      passphrase.upcase.split("").each do |letter|
        shuffle()
        deck.cut(letter.intValue)
      end
    end
  end
  
  def more() 
    shuffle()
    number = deck.cards.first.number
    letter = deck.cards[number].letter
    if letter == nil 
      return more()
    end
    return letter
  end

  def shuffle()
    deck.move(deck.jokerA, 1)
    deck.move(deck.jokerB, 2)
    deck.tripleCut(deck.jokerA, deck.jokerB)
    deck.cut(deck.cards.last.number)
  end
end

class Crypt
  attr_accessor :generator
  
  def initialize(generator)
    @generator = generator
  end
  
  def encrypt(string)
    value = Crypt.encryptable(string)
    keystream = generator.keystream(value.length)
    return value.add(keystream)
  end
  
  def decrypt(string)
    value = string.gsub(" ", "")
    keystream = generator.keystream(value.length)
    return value.subtract(keystream)
  end
  
  def self.encryptable(string)
    value = string.upcase.gsub(/[^A-Z]/i, '')
    while value.length % 5 > 0
      value += "X"
    end
    return value
  end
end

class Test
  attr_accessor :plain
  attr_accessor :pretty
  attr_accessor :encrypted
  attr_accessor :generator
  
  def initialize(plain, pretty, encrypted, generator)
    @plain = plain
    @pretty = pretty
    @encrypted = encrypted
    @generator = generator
  end

  def assert(type, a, b)
    puts type + ": " + (a == b ? 
      "pass (" + a + ")" : 
      "fail (" + a + " / " + b + ")")
  end
  
  def run()
    crypt = Crypt.new(generator.call)
    assert("pretty ", Crypt.encryptable(plain).blocks, pretty)
    assert("encrypt", crypt.encrypt(plain).blocks, encrypted)
    crypt = Crypt.new(generator.call)
    assert("decrypt", crypt.decrypt(encrypted).blocks, pretty)
  end
end

tests = [
  Test.new("Hello!", 
       "HELLO", 
       "IFMMP", 
       Proc.new { Generator.new }),
  Test.new("Do not use PC!", 
       "DONOT USEPC", 
       "OSKJJ JGTMW", 
       Proc.new { Example.new }),
  Test.new("Aaaaaaaaaa!", 
       "AAAAA AAAAA", 
       "EXKYI ZSGEH",
       Proc.new { Solitaire.new(nil) }),
  Test.new("Aaaaaaaaaaaaaaa!", 
       "AAAAA AAAAA AAAAA", 
       "ITHZU JIWGR FARMW", 
       Proc.new { Solitaire.new("FOO") }),
  Test.new("Solitaire!", 
       "SOLIT AIREX", 
       "KIRAK SFJAN", 
       Proc.new { Solitaire.new("CRYPTONOMICON") })
].each { |test| test.run }

File: pontifex/pontifex.py
# This program implements the Solitaire encryption algorithm
# as described in the appendix of the book Cryptonomicon by
# Neal Stephenson:

# https://www.schneier.com/academic/solitaire/

# Solitaire is a cypher that can be performed using a deck of
# playing cards. If two people each have a deck of playing
# cards in the same order, each can generate a keystream that
# can be used to encypher and decypher a message.

from enum import Enum
import string
import textwrap

def charValue(value):
    while value < 1:
        value += 26
    value = (value - 1) % 26 + 1
    return chr(value + 64)

def intValue(char): return ord(char) - 64
def plus(values): return charValue(intValue(values[0]) + intValue(values[1]))
def minus(values): return charValue(intValue(values[0]) - intValue(values[1]))
def split(word): return [char for char in word]
def add(a, b): return "".join(map(plus, zip(split(a), split(b))))
def subtract(a, b): return "".join(map(minus, zip(split(a), split(b))))
def blocks(s): return " ".join(textwrap.wrap(s, 5))

class Suit(Enum):
    CLUBS = 1
    DIAMONDS = 2
    HEARTS = 3
    SPADES = 4
    JOKER = 5

class Card:
    def __init__(self, suit, rank):
        self.suit = suit
        self.rank = rank

    def __str__(self):
        return self.rank_str() + self.suit_str()

    def rank_str(self):
        if self.suit == Suit.JOKER:
            return "A" if self.rank == 1 else "B"
        else:
            letters = {1: "A", 11: "J", 12: "Q", 13: "K"}
            return letters.get(self.rank, str(self.rank))

    def suit_str(self):
        suits = {Suit.CLUBS: "♣️",
                 Suit.DIAMONDS: "♦️",
                 Suit.HEARTS: "♥️",
                 Suit.SPADES: "♠️",
                 Suit.JOKER: "🃏"}
        return suits.get(self.suit)

    def number(self): # subuptimal because we have to calcualte all of these
        suits = {Suit.CLUBS: self.rank,
                 Suit.DIAMONDS: self.rank + 13,
                 Suit.HEARTS: self.rank + 26,
                 Suit.SPADES: self.rank + 39,
                 Suit.JOKER: 53}
        return suits.get(self.suit)

    def letter(self):
        suits = {Suit.CLUBS: charValue(self.rank),
                 Suit.DIAMONDS: charValue(self.rank + 13),
                 Suit.HEARTS: charValue(self.rank),
                 Suit.SPADES: charValue(self.rank + 13),
                 Suit.JOKER: None}
        return suits.get(self.suit)

class Deck:
    def __init__(self):
        self.cards = []
        for suit in [Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS, Suit.SPADES]:
            for rank in range(1, 14):
                self.cards.append(Card(suit, rank))
        self.jokerA = Card(Suit.JOKER, 1)
        self.jokerB = Card(Suit.JOKER, 2)
        self.cards.append(self.jokerA)
        self.cards.append(self.jokerB)

    def __str__(self):
        return ", ".join(map(lambda c: str(c), self.cards))

    def move(self, card, downBy):
        index = self.cards.index(card)
        del self.cards[index]
        newIndex = (index + downBy - 1) % len(self.cards) + 1
        self.cards.insert(newIndex, card)

    def cut(self, count):
        self.cards = (self.cards[count:53] +
                      self.cards[0:count] +
                      self.cards[53:54])

    def tripleCut(self, card1, card2):
        index1 = self.cards.index(card1)
        index2 = self.cards.index(card2)
        if index1 > index2:
            index1, index2 = index2, index1
        self.cards = (self.cards[(index2 + 1):54] +
                      self.cards[index1:(index2 + 1)] +
                      self.cards[0:index1])

class Generator:
    def keystream(self, length):
        stream = ""
        while len(stream) < length:
            stream += self.more()
        return stream[0:length]

    def more(self):
        return "A"

class Example(Generator):
    def more(self):
        return "KDWUPONOWT"

class Solitaire(Generator):
    def __init__(self, passphrase):
        self.deck = Deck()

        if passphrase is not None:
            for letter in split(passphrase.upper()):
                self.shuffle()
                self.deck.cut(intValue(letter))

    def more(self):
        self.shuffle()
        number = self.deck.cards[0].number()
        letter = self.deck.cards[number].letter()
        return letter if letter is not None else self.more()

    def shuffle(self):
        self.deck.move(self.deck.jokerA, 1)
        self.deck.move(self.deck.jokerB, 2)
        self.deck.tripleCut(self.deck.jokerA, self.deck.jokerB)
        self.deck.cut(self.deck.cards[-1].number())

def encryptable(value):
    value = ''.join(c for c in value.upper() if c.isalpha())
    while len(value) % 5 > 0:
        value += "X"
    return value

class Crypt:
    def __init__(self, generator):
        self.generator = generator

    def encrypt(self, value):
        value = encryptable(value)
        keystream = self.generator.keystream(len(value))
        return add(value, keystream)

    def decrypt(self, value):
        value = value.replace(" ", "")
        keystream = self.generator.keystream(len(value))
        return subtract(value, keystream)

class Test:
    def __init__(self, plain, pretty, encrypted, generator):
        self.plain = plain
        self.pretty = pretty
        self.encrypted = encrypted
        self.generator = generator

    def check(self, kind, a, b):
        print(kind + ": " + (
            "pass (" + a + ")" if a == b else
            "fail (" + a + " / " + b + ")"))

    def run(self):
        crypt = Crypt(self.generator())
        self.check("pretty ",
            blocks(encryptable(self.plain)), self.pretty)
        self.check("encrypt",
            blocks(crypt.encrypt(self.plain)), self.encrypted)
        crypt = Crypt(self.generator())
        self.check("decrypt",
            blocks(crypt.decrypt(self.encrypted)), self.pretty)

tests = [Test("Hello!", "HELLO", "IFMMP",
    lambda: Generator()),
  Test("Do not use PC!", "DONOT USEPC", "OSKJJ JGTMW",
    lambda: Example()),
  Test("Aaaaaaaaaa!", "AAAAA AAAAA", "EXKYI ZSGEH",
    lambda: Solitaire(None)),
  Test("Aaaaaaaaaaaaaaa!", "AAAAA AAAAA AAAAA", "ITHZU JIWGR FARMW",
    lambda: Solitaire("FOO")),
  Test("Solitaire!", "SOLIT AIREX", "KIRAK SFJAN",
    lambda: Solitaire("CRYPTONOMICON"))]

for test in tests:
    test.run()